### PR TITLE
Add support of len and index expression for metric

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
Add support for metric values of len expression allowing to get the size of strings, arrays, collections, maps, sets
Add support for metric values of index expression allowing to get element out of arrays, list and maps
Introduced wrapper of ASM type to include generic type information

# Motivation
Support more expressions for metric values

# Additional Notes

### CODE REVIEW

**Review by shatzi — APPROVED** [review]
Very cool!

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/MetricInstrumentor.java [hunk 1]
```diff
@@ -408,7 +485,7 @@ public VisitorResult visit(NumericValue numericValue) {
       InsnList insnList = new InsnList();
       if (number instanceof Long) {
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/MetricInstrumentor.java:486]: No. We agree that we widen all numbers too the largest storage.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/MetricInstrumentor.java [hunk 2]
```diff
@@ -323,7 +332,32 @@ public VisitorResult visit(IsUndefinedExpression isUndefinedExpression) {
...
+        return new VisitorResult(ASMHelper.INT_TYPE, visitorResult.insnList);
+      }
+      if (type.equals(COLLECTION_TYPE)
+          || type.equals(LIST_TYPE)
+          || type.equals(MAP_TYPE)
+          || type.equals(SET_TYPE)) {
+        invokeInterface(visitorResult.insnList, type, "size", Type.INT_TYPE);
+        return new VisitorResult(ASMHelper.INT_TYPE, visitorResult.insnList);
+      }
+      if (type.equals(ARRAYLIST_TYPE)
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/MetricInstrumentor.java:352]: Here we are dealing with Bytecode directly (unlike log/span decoration probes). So we are dealing with static types, not the dynamic ones. So we can't use `instanceof` or alike to determine the type, it's the declared type.

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/MetricProbesInstrumentationTest.java [hunk 3]
```diff
@@ -663,6 +663,89 @@ public void evaluateAtExit() throws IOException, URISyntaxException {
...
+    Assertions.assertTrue(listener.gauges.containsKey(LISTSIZE_METRIC));
+    Assertions.assertEquals(
+        3, listener.gauges.get(LISTSIZE_METRIC).longValue()); // <=> strList.size()
+    Assertions.assertTrue(listener.gauges.containsKey(MAPSIZE_METRIC));
+    Assertions.assertEquals(
+        1, listener.gauges.get(MAPSIZE_METRIC).longValue()); // <=> strMap.size()
+  }
+
+  @Test
+  public void indexExpression() throws IOException, URISyntaxException {
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/MetricProbesInstrumentationTest.java:707]: again, here this is not runtime evaluation, but bytecode execution, so as long as the verifier is not crying, this is runtime error, that we should catch anyway to silent them and not modify user code.
but we can test typing check for better error message

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/MetricProbesInstrumentationTest.java [hunk 4]
```diff
@@ -663,6 +663,89 @@ public void evaluateAtExit() throws IOException, URISyntaxException {
     Assertions.assertEquals(48, listener.counters.get(METRIC_NAME).longValue());
   }
 
+  @Test
+  public void lenExpression() throws IOException, URISyntaxException {
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/MetricProbesInstrumentationTest.java:667]: should just catch all exceptions when executing the call

[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/MetricProbesInstrumentationTest.java:667]: also some part of the Expression Language are not completely implemented like literals

**Pull request merged**